File: Python/src/Wunallet/capaLogica/baseDatos/serializador.py
'''Este módulo fue hecho con la finalidad de crear la función para serilizar los objetos creados, para
así luego poder ser deserializados y usados en el main de nuestra aplicación. Cada tipo de objeto
es serializado por aparte, en su función correspondiente, y luego invocado en la función serializar"""'''

import os
import pathlib
import pickle

from Wunallet.capaLogica.gestorAplicacion.infoClientes.banco import Banco
from Wunallet.capaLogica.gestorAplicacion.infoClientes.perfilCrediticio import PerfilCrediticio
from Wunallet.capaLogica.gestorAplicacion.infoClientes.transaccion import Transaccion
from Wunallet.capaLogica.gestorAplicacion.infoClientes.usuario import Usuario

from Wunallet.capaLogica.gestorAplicacion.productosFinancieros.ahorro import Ahorro
from Wunallet.capaLogica.gestorAplicacion.productosFinancieros.bajoMonto import BajoMonto
from Wunallet.capaLogica.gestorAplicacion.productosFinancieros.corriente import Corriente
from Wunallet.capaLogica.gestorAplicacion.productosFinancieros.credito import Credito

class Serializador():
    
    def serializar(lista, className):
        def camino(className):
            string = os.path.join(pathlib.Path(__file__).parent.absolute(), "temp\\"+className+".txt")
            # string = os.path.join(pathlib.Path(__file__).parent.absolute(), "temp\\"+className)
            print("La ruta es",string)
            return string
        try:
            # Creo el archivo pickle para guardar los objetos
            picklefile = open(camino(className), 'wb')
            # pickle el objeto en el archivo
            pickle.dump(lista, picklefile)
            # cierro el archivo para guardar
            picklefile.close()
            
        except:
            print("paila tuqui tuqui muñeco")

    def serializarTodo():
        print("Estoy aca en serializador, debería funcionar")
        Serializador.serializar(Banco.banco, "Banco")
        Serializador.serializar(PerfilCrediticio._perfilCrediticio, "PerfilCreditico")
        Serializador.serializar(Transaccion.getTransaccion(), "Transaccion")
        Serializador.serializar(Usuario.getUsuario(), "Usuario")
        Serializador.serializar(Ahorro.getAhorro(), "Ahorro")
        Serializador.serializar(BajoMonto.getBajoMonto(), "BajoMonto")
        Serializador.serializar(Corriente.getCorriente(), "Corriente")
        Serializador.serializar(Credito.getCredito(), "Credito")
        # Serializador.serializar(Empleado.getEmpleados(), "Empleados")

        # print("Banco:", Banco.banco)
        # print("PerfilCrediticio:", PerfilCrediticio._perfilCrediticio)
        # print("Transaccion:", Transaccion.getTransaccion())
        # print("Usuario:", Usuario.getUsuario())
        # print("Ahorro:", Ahorro.getAhorro())
        # print("BajoMonto:", BajoMonto.getBajoMonto())
        # print("Corriente:", Corriente.getCorriente())
        # print("Credito:", Credito.getCredito())

File: Java/src/basedatos/Serializador.java
package basedatos;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.util.List;

import gestorAplicacion.infoClientes.Banco;
import gestorAplicacion.infoClientes.PerfilCreditico;
import gestorAplicacion.infoClientes.Transaccion;
import gestorAplicacion.infoClientes.Usuario;

import gestorAplicacion.productosFinancieros.Ahorro;
import gestorAplicacion.productosFinancieros.BajoMonto;
import gestorAplicacion.productosFinancieros.Corriente;
import gestorAplicacion.productosFinancieros.Credito;


/**
 * Se utiliza para serializar todos los objetos creados durante la ejecucion
 * del proyecto
 * @author Erik Gonzalez
 * @author Felipe Miranda
 */
public class Serializador {
	/**
	 * Serializamos una lista por el nombre de la clase
	 * 
	 * @param <E>       el generico se usa para poder agredar las clases que se
	 *                  crearon
	 * @param lista     Una lista de objetos
	 * @param className El nombre de la clase que queremos usar como nombre del
	 *                  archivo
	 */
	public static <E> void serializar(List<E> lista, String className) {
		FileOutputStream fileOut;

		try {
			String path = System.getProperty("user.dir") + "/src/baseDatos/temp/" + className + ".txt";
			// se crea un fileoutputstream para saber donde serializar los archivos
			fileOut = new FileOutputStream(path);
			// Se crea un objeto output stream para poder escribir en el archivo
			ObjectOutputStream out = new ObjectOutputStream(fileOut);
			// Guardamos la lista de objetos
			out.writeObject(lista);
			out.close();
			fileOut.close();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}

	/**
	 * Serializamos todas las clases que necesitamos
	 */
	public static void serializarTodo() {
		Serializador.serializar(Banco.listaBancos, "Banco");
		Serializador.serializar(PerfilCreditico.getPerfilCreditico(), "PerfilCreditico");
		Serializador.serializar(Transaccion.getTransaccion(), "Transaccion");
		Serializador.serializar(Usuario.getUsuario(), "Usuario");
		Serializador.serializar(Ahorro.getAhorro(), "Ahorro");
		Serializador.serializar(BajoMonto.getBajoMonto(), "BajoMonto");
		Serializador.serializar(Corriente.getCorriente(), "Corriente");
		Serializador.serializar(Credito.getCredito(), "Credito");
	}
}

